src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LspFileBasedProgramWorkspace.cs

```diff
@@ -0,0 +1,323 @@
+        var parseOptions = languageServices.GetService<ISyntaxTreeFactoryService>()?.GetDefaultParseOptionsWithLatestLanguageVersion();
+
+        // It seems like the workflow in main branch is:
+        // 1. add file to misc workspace
+        // 2. process the project in background
+        // 3. realize file is part of the project
+        // 4. remove from misc workspace and reopen in host workspace
+        // It might be expensive if (1) effectively changes from "just open in misc workspace" versus "generate/process it as its own project etc.".
+        // We might need a stronger heuristic that the current file is unlikely to be part of a project. For example, that there is no .csproj in any of the containing directories.
+        if (!checkIsFileBasedProgram())
+        {
+            return null;
+        }
+
+        using (await _gate.DisposableWaitAsync(cancellationToken))
+        {
+            _projectsToLoadAndReload.AddWork((uri, documentText, languageId));
+
+            // Wait for the in progress batch to complete and send a project initialized notification to the client.
+            await _projectsToLoadAndReload.WaitUntilCurrentBatchCompletesAsync();
+            var documentId = this.CurrentSolution.GetDocumentIdsWithFilePath(documentFilePath).First();
+            var document = this.CurrentSolution.GetDocument(documentId);
+            return document;
+        }
+
+        bool checkIsFileBasedProgram()
+        {
+            var languageServices = Services.SolutionServices.GetLanguageServices(languageInformation.LanguageName);
+            if (parseOptions is not CSharpParseOptions csharpOptions)
+                return false;
+
+            var tree = CSharpSyntaxTree.ParseText(text: documentText, options: csharpOptions, path: documentFilePath);
+            var root = tree.GetRoot();
+            return
+                //root.DescendantNodes(descendIntoChildren: node => node.IsKind(SyntaxKind.CompilationUnit)).OfType<GlobalStatementSyntax>().Any() ||
+                root.GetLeadingTrivia().FirstOrDefault().IsKind(SyntaxKind.ShebangDirectiveTrivia);
+        }
+    }
+
+    private async ValueTask LoadOrReloadProjectsAsync(ImmutableSegmentedList<(Uri uri, SourceText documentText, string languageId)> list, CancellationToken token)
+    {
+        foreach (var pair in list)
+        {
+            await LoadOrReloadProjectAsync(pair, logger: null, token);
+        }
+    }
+
+    /// <summary>
+    /// Takes in a file URI and text and creates a misc project and document for the file.
+    /// 
+    /// Calls to this method and <see cref="TryRemoveMiscellaneousDocument(Uri, bool)"/> are made
+    /// from LSP text sync request handling which do not run concurrently.
+    /// </summary>
+    public async Task<Document?> LoadOrReloadProjectAsync((Uri uri, SourceText documentText, string languageId) pair, ILspLogger? logger, CancellationToken cancellationToken)
```

Comment:
this seems very weird that the ABWQ calls this *and* it's public.  The reason for ABWQ is for serialized ordering of work.  having this callable concurrently from both is an anti pattern.